### DISCUSSION TITLE
How do you install NVM in a Windows container?

### DISCUSSION THREAD

**Discussion by totszwai** [coreybutler/nvm-windows#1175] [Support]
```
FROM mcr.microsoft.com/windows/server:ltsc2022
SHELL ["powershell", "-Command", "$ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue';"]

USER ContainerAdministrator

RUN net accounts /maxpwage:unlimited
RUN net user root /add /active:yes /homedir:C:\Users\root /expires:never
RUN net localgroup administrators /add root
RUN wmic USERACCOUNT set PasswordExpires=False

RUN setx /M NVM_HOME 'C:\Users\root\.nvm'
RUN setx /M NODE_VERSION 17

RUN curl -o nvm.zip https://github.com/coreybutler/nvm-windows/releases/download/1.1.12/nvm-noinstall.zip
RUN mkdir -p ${Env:NVM_HOME}
RUN tar -xf nvm.zip -C ${Env:NVM_HOME}

# If I don't create this file, it will complaint that the \settings.txt not found
RUN New-Item -Path "$Env:NVM_HOME\settings.txt" -ItemType "File"
```

If I try to install any version, it will complain that it doesn't know where to save the npm. This is inside the Windows container.

```
C:\Users\root\.nvm>nvm install 17
 could not be found or does not exist. Exiting.
Downloading node.js version 17.9.1 (64-bit)... 
Error while creating \v17.9.1\node.zip - open \v17.9.1\node.zip: The system cannot find the path specified.
Could not download node.js v17.9.1 64-bit executable.
```

I had to run this manually for it to work `nvm root C:\Users\root\.nvm`, it will then add a brunch of stuff into `settings.txt`, however, I **cannot** run this step in my Dockerfile, it would just hang during build phase.

I also tried creating the settings.txt manually like this, but then `nvm` still won't "sees" the settings, and would throw the exact same errors when I try to do `nvm install 17`.
```
# RUN echo `
# root: C:\Users\root\.nvm `
# arch: 64 `
# proxy: none `
# originalpath: `
# originalversion: `
# node_mirror: `
# npm_mirror: `
# > "$Env:NVM_HOME\settings.txt"
```

I also tried running the installer in the Dockerfile, but it just won't do anything. Running it inside the Container won't do anything either.

```
RUN curl -o nvm-setup.exe https://github.com/coreybutler/nvm-windows/releases/download/1.1.12/nvm-setup.exe
RUN nvm-setup.exe
```

So how can pre-install `nvm` in my Windows container??

**Comment by coreybutler** [comment]
Did you set `NVM_SYMLINK` anywhere? I did not see that in your Dockerfile. When you set it, the directory should **not** exist. For example, `RUN setx /M NVM_SYMLINK C:\Program Files\nodejs` (where `nodejs` does not exist). Additionally, `NVM_HOME` and `NVM_SYMLINK` both need to be added to the `PATH`. There is a guide at https://github.com/coreybutler/nvm-windows/wiki#manual-installation.

Your second approach (running `nvm-setup.exe`) won't work in a container because it is a GUI installer.
